mernarezk/assignment2/Q1.java

```diff
@@ -0,0 +1,114 @@
+package Assignment2;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.swing.plaf.synth.SynthSpinnerUI;
+
+public class Q1 {
+	/**
+	 * Method is divided into two parts, the first part calls searchTarget method which search all nodes
+	 * of the binary tree until the current node matches our target node and returns it, the second part
+	 * just travels this target node's parent up till the root and adds them to a list.
+	 * @param root root of the tree
+	 * @param target target value that we need to get it's parents
+	 * @return a list of the target node's parents.
+	 * @throws IllegalStateException
+	 */
+	public static List<Integer> getAncestorsWithParent(TreeNode root, int target) throws IllegalStateException {
+		TreeNode result = searchTarget(root, target);
+		if (result == null) {
+			throw new IllegalStateException("Target was not found in the tree.");
+		}
+		List<Integer> ancestors = new ArrayList<Integer>();
+		result = result.parent;
+		while (result != null) {
+			ancestors.add(0, result.val);
+			result = result.parent;
+		}
+		return ancestors;
+	}
+
+	private static TreeNode searchTarget(TreeNode root, int target) {
+		TreeNode result = null;
+		if (root == null)
+			return null;
+		if (root.val == target)
+			result = root;
+		if (root.left != null)
+			result = searchTarget(root.left, target);
+		if (root.right != null)

```

Comment:
What if the root has both left and right children, and the target is found in the left tree and not in the right tree? The second call to searchTarget would override the result. I think it should be enough to add if (result != null) to these 'if' statements too.

Reply:
Oh yes that's true. fixed